File: src/main/java/de/twometer/amongus/model/Player.java
package de.twometer.amongus.model;

import org.joml.Vector3f;

import java.util.List;

public class Player implements PlayerBehavior {

    public int id;

    public String username;

    public Vector3f position;

    public float rotation;

    public PlayerRole role;

    public PlayerColor color;

    public List<PlayerTask> tasks;

    public boolean alive;

    public int emergencyMeetings = 0;

    public int getId() {
        return id;
    }

    public String getUsername() {
        return username;
    }

    public PlayerRole getRole() {
        return role;
    }

    public PlayerColor getColor() {
        return color;
    }

    public boolean canDoTask(Location location, TaskType taskType) {
        if (role == PlayerRole.Impostor) return false;
        return findRunningTaskByStage(location, taskType) != null;
    }

    public PlayerTask findRunningTaskByStage(Location location, TaskType taskType) {
        for (var task : tasks) {
            var stage = task.getNextStage();
            if (stage.getTaskType() == taskType && stage.getLocation() == location && !task.isCompleted())
                return task;
        }
        return null;
    }

}

File: src/main/kotlin/de/twometer/amogus/model/Player.kt
package de.twometer.amogus.model

import de.twometer.neko.scene.nodes.ModelNode
import org.joml.Vector3f

class Player : IPlayer {
    override var id = IPlayer.INVALID_PLAYER_ID
    override var username = ""
    override var role: PlayerRole = PlayerRole.Crewmate
    override var color: PlayerColor = PlayerColor.Red
    val prevPosition: Vector3f = Vector3f(0f, 0f, 0f)
    val position: Vector3f = Vector3f(0f, 0f, 0f)
    var prevRotation = 0f
    var rotation = 0f
    var lastUpdate: Long = 0L
    val speedSquared get() = position.distanceSquared(prevPosition)
    override var state: PlayerState = PlayerState.Alive
    var emergencyMeetings = 0
    var lastMeetingCalled = 0L
    var tasks: List<PlayerTask> = ArrayList()
    var node: ModelNode? = null
    var killCooldown: Int = 0

    fun findTask(location: Location, type: TaskType): PlayerTask? = tasks.firstOrNull {
        val nextStage = it.nextStage
        !it.isCompleted && !it.isTimerRunning && nextStage.location == location && nextStage.taskType == type
    }
}